crates/utils/src/rate_limit/backend.rs

```diff
@@ -0,0 +1,302 @@
+//! The content in this file is mostly copy-pasted from library code:
+//! https://github.com/jacob-pro/actix-extensible-rate-limit/blob/master/src/backend/memory.rs
+
+use crate::rate_limit::{
+  ActionType,
+  BucketConfig,
+  input::{LemmyInput, RateLimitIpAddr},
+};
+use actix_extensible_rate_limit::backend::{
+  Backend,
+  Decision,
+  SimpleOutput,
+  memory::DEFAULT_GC_INTERVAL_SECONDS,
+};
+use actix_web::rt::{task::JoinHandle, time::Instant};
+use dashmap::DashMap;
+use enum_map::EnumMap;
+use std::{
+  convert::Infallible,
+  sync::{Arc, RwLock},
+  time::Duration,
+};
+
+/// A Fixed Window rate limiter [Backend] that uses [Dashmap](dashmap::DashMap) to store keys
+/// in memory.
+#[derive(Clone)]
+pub struct LemmyBackend {
+  map: Arc<DashMap<RateLimitIpAddr, Value>>,
+  gc_handle: Option<Arc<JoinHandle<()>>>,
+  pub(super) configs: Arc<RwLock<EnumMap<ActionType, BucketConfig>>>,
+}
+
+struct Value {
+  ttl: Instant,
+  count: u64,
+}
+
+impl LemmyBackend {
+  pub(crate) fn new(configs: EnumMap<ActionType, BucketConfig>, enable_gc: bool) -> Self {
+    let map = Arc::new(DashMap::<RateLimitIpAddr, Value>::new());
+    let gc_handle = enable_gc.then(|| {
+      Arc::new(LemmyBackend::garbage_collector(
+        map.clone(),
+        Duration::from_secs(DEFAULT_GC_INTERVAL_SECONDS),
+      ))
+    });
+    LemmyBackend {
+      map,
+      gc_handle,
+      configs: Arc::new(RwLock::new(configs)),
+    }
+  }
+
+  fn garbage_collector(
+    map: Arc<DashMap<RateLimitIpAddr, Value>>,
+    interval: Duration,
+  ) -> JoinHandle<()> {
+    assert!(
+      interval.as_secs_f64() > 0f64,
+      "GC interval must be non-zero"
+    );
+    tokio::spawn(async move {
+      loop {
+        let now = Instant::now();
+        map.retain(|_k, v| v.ttl > now);
+        tokio::time::sleep_until(now + interval).await;
+      }
+    })
+  }
+}
+
+impl Backend<LemmyInput> for LemmyBackend {
+  type Output = SimpleOutput;
+  type RollbackToken = RateLimitIpAddr;
+  type Error = Infallible;
+
+  #[expect(clippy::expect_used)]
+  async fn request(
+    &self,
+    input: LemmyInput,
+  ) -> Result<(Decision, Self::Output, Self::RollbackToken), Self::Error> {
+    #[allow(clippy::expect_used)]
+    let config = self.configs.read().expect("read rwlock")[input.action_type];
+
+    let max_requests: u64 = config.max_requests.into();
+    let interval = Duration::from_secs(config.interval.into());
+
+    let now = Instant::now();
+    let mut count = 1;
+    let mut expiry = now
+      .checked_add(interval)
+      .expect("Interval unexpectedly large");
+    self
+      .map
+      .entry(input.key)
```

Comment:
won't the key need to include the action type? Otherwise you would end up putting different limits under the same counter?